sql/src/main/java/org/apache/druid/sql/calcite/schema/BrokerSegmentMetadataCache.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    synchronized (lock) {
+      // Add missing segments back to the refresh list.
+      segmentsNeedingRefresh.addAll(Sets.difference(segmentsToRefresh, refreshed));
+
+      // Compute the list of datasources to rebuild tables for.
+      dataSourcesToRebuild.addAll(dataSourcesNeedingRebuild);
+      refreshed.forEach(segment -> dataSourcesToRebuild.add(segment.getDataSource()));
+
+      // Remove those datasource for which we received schema from the Coordinator.
+      dataSourcesToRebuild.removeAll(polledDataSourceMetadata.keySet());
+      dataSourcesNeedingRebuild.clear();
+    }
+
+    // Rebuild the datasources.
+    for (String dataSource : dataSourcesToRebuild) {
+      final RowSignature rowSignature = buildDataSourceRowSignature(dataSource);
+      if (rowSignature == null) {
+        log.info("datasource [%s] no longer exists, all metadata removed.", dataSource);
+        tables.remove(dataSource);
+        return;
+      }
+
+      final PhysicalDatasourceMetadata physicalDatasourceMetadata = dataSourceMetadataFactory.build(dataSource, rowSignature);
+      updateDSMetadata(dataSource, physicalDatasourceMetadata);
+    }
+  }
+
+  private Map<String, PhysicalDatasourceMetadata> queryDataSourceInformation(Set<String> dataSourcesToQuery)
+  {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+
+    List<DataSourceInformation> dataSourceInformations = null;
+
+    emitter.emit(ServiceMetricEvent.builder().setMetric(
+        "metadatacache/schemaPoll/count", 1));
+    try {
+      dataSourceInformations = FutureUtils.getUnchecked(coordinatorClient.fetchDataSourceInformation(dataSourcesToQuery), true);
+    }
+    catch (Exception e) {
+      log.debug(e, "Failed to query datasource information from the Coordinator.");
+      emitter.emit(ServiceMetricEvent.builder().setMetric(
+          "metadatacache/schemaPoll/failed", 1));
+    }
+
+    emitter.emit(ServiceMetricEvent.builder().setMetric(
+        "metadatacache/schemaPoll/time",
+        stopwatch.elapsed(TimeUnit.MILLISECONDS)));
+
+    final Map<String, PhysicalDatasourceMetadata> polledDataSourceMetadata = new HashMap<>();
+
+    if (dataSourceInformations != null) {
+      dataSourceInformations.forEach(dataSourceInformation -> polledDataSourceMetadata.put(
+          dataSourceInformation.getDataSource(),
+          dataSourceMetadataFactory.build(
+              dataSourceInformation.getDataSource(),
+              dataSourceInformation.getRowSignature()
+          )
+      ));
+    }
+
+    return polledDataSourceMetadata;
+  }
+
+  private void updateDSMetadata(String dataSource, PhysicalDatasourceMetadata physicalDatasourceMetadata)
+  {
+    final PhysicalDatasourceMetadata oldTable = tables.put(dataSource, physicalDatasourceMetadata);
+    if (oldTable == null || !oldTable.getRowSignature().equals(physicalDatasourceMetadata.getRowSignature())) {
+      log.info("[%s] has new signature: %s.", dataSource, physicalDatasourceMetadata.getRowSignature());
```

Comment:
I haven't quite figured it out yet, or if it is related to the refactor in this PR or some pre-existing problem, but might be worth looking further into anyway... 

while bouncing a historical repeatedly, so that datasources are going offline and online periodically, i keep finding log messages where it gets into a state that it is printing an empty row signature.
```
2023-10-28T03:33:46,084 INFO [DruidSchema-Cache-0] org.apache.druid.sql.calcite.schema.BrokerSegmentMetadataCache - [wikipedia-schemaless] has new signature: {}.
```
which I think must mean that it knows the table needs refreshed, but has not yet obtained any segment metadata. I'll try to see if i can determine how this happens

Reply:
When the historical went offline, the table would have been removed from the cache, when it again came up, the datasource schema is built and the message is logged since this condition `oldTable == null` is met. 

`RowSignature` being empty implies that SMQ queries succeeded but returned empty `rowSignature` for the segments? is that possible?
Not sure if this got introduced by the new changes, since we are just fetching the schema from Coordinator (which is having empty `rowSignature`) and the flow on Coordinator is same as what used to be on the Broker.